### DISCUSSION TITLE
How to disable automatic removal of finished triggers

### DISCUSSION THREAD

**Discussion by mrtzd** [quartznet/quartznet#1703] [Q&A] [answered]
Right now every trigger is automatically removed after it is fired for the last time. Is there any way that I can keep the triggers in the database even when they're not going to be fired again?

**Comment by mrtzd** [comment]
I looked into this a bit more. At first it seemed like that I can solve this problem by implementing my own TriggerListener and overriding the TriggerComplete method. 
```
public override Task TriggerComplete(ITrigger trigger, IJobExecutionContext context, SchedulerInstruction triggerInstructionCode, CancellationToken cancellationToken = default)
    {
        return base.TriggerComplete(trigger, context, triggerInstructionCode, cancellationToken);
    }
```
Unfortunately it seems like this interface is just for monitoring purposes and I can't actually change the way triggers are handled internally (changing the value of 'triggerInstructionCode' doesn't have any effect). 
I'm not sure if I'm missing something here or not so any help would be really appreciated. Thank you

**Comment by lahma** [comment, marked as answer]
In 10 years time I can't remember hearing about this kind of need. What's you use case? You could probably use `ISchedulerListener.TriggerFinalized` to record trigger data to some history table or something similar.

**Reply by mrtzd** [reply]
> In 10 years time I can't remember hearing about this kind of need. What's you use case?

Basically I have created a web interface and through this interface users can create their own tasks (implemented as one durable job with many triggers). I wanted to have the option so that users can see their expired/completed tasks and have the option to reschedule them instead of creating new ones from scratch.

> You could probably use ISchedulerListener.TriggerFinalized to record trigger data to some history table or something similar.

Thank you very much for the suggestion, I think I'll do something like that. It wouldn't cause any issues with the scheduler if I created my own tables inside Quartz database?

**Reply by lahma** [reply]
As long as you don't touch Quartz tables you should be fine. What you described sounds like trigger history (there's a plugin sample in source repo) and then you would also record the original configuration. Supporting multiple trigger types will cause a bit of repetition though - maybe you should model the trigger snapshot data based on your UI scheduling capabilities / limit possibilities.

**Reply by CrazyMoka** [reply]
I have a situation where triggers state is COMPLETE but they are not removed. When would this happen?
I did notice that missfire hit 4 times which is weird because it was set to only run instantly with no repeat.
Please advise.

**Reply by CrazyMoka** [reply]
Actually the trigger is STATE complete but the job is still running.
